fontbe/src/mvar.rs

```diff
@@ -0,0 +1,329 @@
+
+    fn add_sources(&mut self, mvar_tag: Tag, sources: &GlobalMetricValues) -> Result<(), Error> {
+        let sources: HashMap<_, _> = sources
+            .iter()
+            .map(|(loc, src)| (loc.clone(), vec![src.into_inner() as f64]))
+            .collect();
+        if sources.len() == 1 {
+            assert!(sources.keys().next().unwrap().is_default());
+            // spare the model the work of computing no-op deltas
+            return Ok(());
+        }
+        let locations = sources.keys().cloned().collect::<BTreeSet<_>>();
+        let model = self.models.entry(locations).or_insert_with(|| {
+            // this glyph defines its own set of locations, a new sparse model is needed
+            VariationModel::new(sources.keys().cloned().collect(), self.axes.clone()).unwrap()
+        });
+        let deltas: Vec<_> = model
+            .deltas(&sources)
+            .map_err(|e| Error::MvarDeltaError(mvar_tag, e))?
+            .into_iter()
+            .filter_map(|(region, values)| {
+                if region.is_default() {
+                    return None;
+                }
+                // Only 1 value per region for our input
+                assert!(values.len() == 1, "{} values?!", values.len());
+                Some((
+                    region.to_write_fonts_variation_region(&self.axes),
+                    values[0].ot_round(),
+                ))
+            })
+            .collect();
+        // don't encode no-op deltas
+        if deltas.iter().all(|(_, delta)| *delta == 0) {
+            return Ok(());
+        }
+        self.deltas.insert(mvar_tag, deltas);
+        Ok(())
+    }
+
+    fn build(self) -> Option<Mvar> {
+        let mut builder = VariationStoreBuilder::new(self.axes.len() as u16);
+        let delta_ids = self
+            .deltas
+            .into_iter()
+            .map(|(tag, deltas)| (tag, builder.add_deltas(deltas)))
+            .collect::<Vec<_>>();
+
+        let (varstore, index_map) = builder.build();
+
+        let records = delta_ids
+            .into_iter()
+            .map(|(tag, temp_id)| {
+                let varidx = index_map.get(temp_id).unwrap();
+                ValueRecord::new(
+                    tag,
+                    varidx.delta_set_outer_index,
+                    varidx.delta_set_inner_index,
+                )
+            })
+            .collect::<Vec<_>>();
+
+        (!records.is_empty()).then(|| Mvar::new(MajorMinor::VERSION_1_0, Some(varstore), records))
+    }
+}
+
+impl Work<Context, AnyWorkId, Error> for MvarWork {
+    fn id(&self) -> AnyWorkId {
+        WorkId::Mvar.into()
+    }
+
+    fn read_access(&self) -> Access<AnyWorkId> {
+        AccessBuilder::new()
+            .variant(FeWorkId::StaticMetadata)
+            .variant(FeWorkId::GlobalMetrics)
+            .build()
+    }
+
+    /// Generate [MVAR](https://learn.microsoft.com/en-us/typography/opentype/spec/MVAR)
+    fn exec(&self, context: &Context) -> Result<(), Error> {
```

Comment:
Optional, is there corresponding python code worth referencing?

Reply:
Done